### ISSUE TITLE
Remove S3 secrets from logs/environment variables

### ISSUE THREAD

**Issue by dvianello** [CrunchyData/postgres-operator#1306]
**What is the motivation or use case for the change? **
S3 secrets for Backrest backups are visible in:
- environment variables
- logs

**Describe the solution you'd like**
- Secrets not visible in logs/environment variables
- Secrets mounted via K8S secrets

**Please tell us about your environment:**
- Operating System: CentOS
- Where is this running ( Local, Cloud Provider): VMware
- Storage being used (NFS, Hostpath, Gluster, etc): VMware vSan
- Container Image Tag: `v1.16.3`
- PostgreSQL Version: `centos7-12.1-4.2.1`
- Platform (Docker, Kubernetes, OpenShift): Kubernetes
- Platform Version: `v1.16.3`

**Comment by jkatz** [comment]
@dvianello I agree 100% with this. Do you have redacted log examples with line info? That will help with tracing it. Thanks!

**Comment by jkatz** [comment]
Also please provide the information that is requested in the bug reporting template:

* Operating System:
* Where is this running ( Local, Cloud Provider)
* Storage being used (NFS, Hostpath, Gluster, etc):
* Container Image Tag:
* PostgreSQL Version:
* Platform (Docker, Kubernetes, OpenShift):
* Platform Version:

**Comment by dvianello** [comment]
@jkatz, updated the initial post above including the info you required.

Going back to where secrets are visible, if I configure the Operator (deployed via Ansible) with S3 secrets, and then spin up a cluster with
```pgo create cluster test  --pgbackrest-storage-type=local,s3 -n postgres-dbs```

I can see secrets in clear-text in the master description:
```
Name:         test-6874bb948b-7hntp
Namespace:    postgres-dbs
Priority:     0
Node:         node5/xxx.xxx.xxx.xxx
Start Time:   Mon, 02 Mar 2020 15:05:28 +0000
Labels:       archive-timeout=60
              crunchy-pgha-scope=test
              crunchy_collect=true
              deployment-name=test
              name=test
              pg-cluster=test
              pg-cluster-id=250b36ab-8310-44f5-87f2-bae9de7832fd
              pg-pod-anti-affinity=
              pgo-pg-database=true
              pgo-version=4.2.1
              pgouser=admin
              pod-template-hash=6874bb948b
              role=master
              service-name=test
              vendor=crunchydata
              workflowid=0ab6d073-39b6-4359-967a-85bf5df77ded
Annotations:  status:
                {"conn_url":"postgres://xxx.xxx.xxx.xxx:5432/postgres","api_url":"http://xxx.xxx.xxx.xxx:8009/patroni","state":"running","role":"master","versio...
Status:       Running
IP:           xxx.xxx.xxx.xxx
IPs:
  IP:           xxx.xxx.xxx.xxx
Controlled By:  ReplicaSet/test-6874bb948b
Containers:
  database:
    Container ID:   docker://e11ce8512e7219004b6885ce307dafba214fa79d6ca9c6d3a606c958ea75f28d
    Image:          crunchydata/crunchy-postgres-ha:centos7-12.1-4.2.1
    Image ID:       docker-pullable://crunchydata/crunchy-postgres-ha@sha256:d307b86ee7ebc6b288b6cd11726366a71f0a163caa041eecd4aee78f40cc1154
    Ports:          5432/TCP, 8009/TCP
    Host Ports:     0/TCP, 0/TCP
    State:          Running
      Started:      Mon, 02 Mar 2020 15:05:40 +0000
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     4
      memory:  4Gi
    Requests:
      cpu:      2
      memory:   2Gi
    Liveness:   exec [/opt/cpm/bin/pgha-liveness.sh] delay=30s timeout=10s period=15s #success=1 #failure=3
    Readiness:  exec [/opt/cpm/bin/pgha-readiness.sh] delay=15s timeout=1s period=10s #success=1 #failure=3
    Environment:
      PGHA_PG_PORT:                       5432
      PGHA_USER:                          postgres
      PGHA_INIT:                          <set to the key 'init' of config map 'test-pgha-default-config'>  Optional: false
      PATRONI_POSTGRESQL_DATA_DIR:        /pgdata/test
      PGMONITOR_PASSWORD:                 password
      PGBACKREST_STANZA:                  db
      PGBACKREST_REPO1_HOST:              test-backrest-shared-repo
      BACKREST_SKIP_CREATE_STANZA:        true
      PGHA_PGBACKREST:                    true
      PGBACKREST_REPO1_PATH:              /backrestrepo/test-backrest-shared-repo
      PGBACKREST_DB_PATH:                 /pgdata/test
      ENABLE_SSHD:                        true
      PGBACKREST_LOG_PATH:                /tmp
      PGBACKREST_PG1_SOCKET_PATH:         /tmp
      PGBACKREST_PG1_PORT:                5432
      PGBACKREST_REPO_TYPE:               posix
      PGHA_PGBACKREST_LOCAL_S3_STORAGE:   true
      PGBACKREST_REPO1_S3_BUCKET:         pgo-backups
      PGBACKREST_REPO1_S3_ENDPOINT:       xxx.test.com
      PGBACKREST_REPO1_S3_REGION:         eu-west-1
      PGBACKREST_REPO1_S3_KEY:            pgo
      *PGBACKREST_REPO1_S3_KEY_SECRET*:     xxxxxxx
      PGBACKREST_REPO1_S3_CA_FILE:        /sshd/aws-s3-ca.crt
      PGBACKREST_REPO1_HOST_CMD:          /usr/local/bin/archive-push-s3.sh
      PGHA_DATABASE:                      database
      PGHA_CRUNCHYADM:                    true
      PGHA_REPLICA_REINIT_ON_START_FAIL:  true
      PGHA_SYNC_REPLICATION:              false
      PATRONI_KUBERNETES_NAMESPACE:       postgres-dbs (v1:metadata.namespace)
      PATRONI_KUBERNETES_SCOPE_LABEL:     crunchy-pgha-scope
      PATRONI_SCOPE:                       (v1:metadata.labels['crunchy-pgha-scope'])
      PATRONI_KUBERNETES_LABELS:          {vendor: "crunchydata"}
      PATRONI_LOG_LEVEL:                  INFO
      PGHOST:                             /tmp
    Mounts:
      /backrestrepo from backrestrepo (rw)
      /crunchyadm from crunchyadm (rw)
      /pgconf from pgconf-volume (rw)
      /pgconf/pgreplicator from primary-volume (rw)
      /pgconf/pgsuper from root-volume (rw)
      /pgconf/pguser from user-volume (rw)
      /pgdata from pgdata (rw)
      /pgwal from pgwal-volume (rw)
      /recover from recover-volume (rw)
      /sshd from sshd (ro)
      /var/run/secrets/kubernetes.io/serviceaccount from pgo-pg-token-554rr (ro)
  collect:
    Container ID:   docker://9c7dc3b1e6eb6302037674b42bbdc65cd5d5b34acb697aa73c90491731bbff0b
    Image:          crunchydata/crunchy-collect:centos7-12.1-4.2.1
    Image ID:       docker-pullable://crunchydata/crunchy-collect@sha256:91176aeedf147d939d58edd5c48b8fe00fd48c3d875c06d6e4de654cd339cde7
    Port:           9187/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Mon, 02 Mar 2020 15:05:40 +0000
    Ready:          True
    Restart Count:  0
    Environment:
      COLLECT_PG_HOST:         127.0.0.1
      COLLECT_PG_PORT:         5432
      COLLECT_PG_DATABASE:     postgres
      COLLECT_PG_PARAMS:       sslmode=disable
      JOB_NAME:                test
      POSTGRES_EXPORTER_PORT:  9187
    Mounts:
      /collect-pguser from collect-volume (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from pgo-pg-token-554rr (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  pgdata:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  test
    ReadOnly:   false
  user-volume:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  test-user-secret
    Optional:    false
  primary-volume:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  test-primaryuser-secret
    Optional:    false
  collect-volume:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  test-collect-secret
    Optional:    false
  sshd:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  test-backrest-repo-config
    Optional:    false
  root-volume:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  test-postgres-secret
    Optional:    false
  pgwal-volume:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  recover-volume:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  report:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  backrestrepo:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:     Memory
    SizeLimit:  <unset>
  crunchyadm:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  pgconf-volume:
    Type:               Projected (a volume that contains injected data from multiple sources)
    ConfigMapName:      test-pgha-default-config
    ConfigMapOptional:  0xc0007f4409
  pgo-pg-token-554rr:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  pgo-pg-token-554rr
    Optional:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

I can also see secrets in clear-text in the pod description for the `backrest` repo
```
Name:         test-backrest-shared-repo-b5645cbdd-n5tq7
Namespace:    postgres-dbs
Priority:     0
Node:         node4/xxx.xxx.xxx.xxx
Start Time:   Mon, 02 Mar 2020 15:05:28 +0000
Labels:       name=test-backrest-shared-repo
              pg-cluster=test
              pg-pod-anti-affinity=
              pgo-backrest-repo=true
              pod-template-hash=b5645cbdd
              service-name=test-backrest-shared-repo
              vendor=crunchydata
Annotations:  <none>
Status:       Running
IP:           xxx.xxx.xxx.xxx
IPs:
  IP:           xx.xxx.xxx.xxx
Controlled By:  ReplicaSet/test-backrest-shared-repo-b5645cbdd
Containers:
  database:
    Container ID:   docker://XXXX
    Image:          crunchydata/pgo-backrest-repo:centos7-4.2.1
    Image ID:       docker-pullable://crunchydata/pgo-backrest-repo@sha256:2d91d5e567f2266245386abf043da07b270e6aebcf80dada0001280347bcb221
    Port:           2022/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Mon, 02 Mar 2020 15:05:47 +0000
    Ready:          True
    Restart Count:  0
    Environment:
      PGBACKREST_REPO1_S3_BUCKET:      pgo-backups
      PGBACKREST_REPO1_S3_ENDPOINT:    xxx.test.com
      PGBACKREST_REPO1_S3_REGION:      eu-west-1
      PGBACKREST_REPO1_S3_KEY:         pgo
      *PGBACKREST_REPO1_S3_KEY_SECRET:*  xxxxxxxxxxxxxxxxxxxxxxxxxxxxx
      PGBACKREST_REPO1_S3_CA_FILE:     /sshd/aws-s3-ca.crt
      PGBACKREST_REPO1_HOST_CMD:       /usr/local/bin/archive-push-s3.sh
      PGBACKREST_STANZA:               db
      SSHD_PORT:                       2022
      PGBACKREST_DB_PATH:              /pgdata/test
      PGBACKREST_REPO_PATH:            /backrestrepo/test-backrest-shared-repo
      PGBACKREST_PG1_PORT:             5432
      PGBACKREST_LOG_PATH:             /tmp
      PGBACKREST_PG1_SOCKET_PATH:      /tmp
      PGBACKREST_DB_HOST:              test
    Mounts:
      /backrestrepo from backrestrepo (rw)
      /sshd from sshd (ro)
```

and in its logs:
```
PGBACKREST env vars are set to:
PGBACKREST_DB_HOST=test
PGBACKREST_DB_PATH=/pgdata/test
PGBACKREST_LOG_PATH=/tmp
PGBACKREST_PG1_PORT=5432
PGBACKREST_PG1_SOCKET_PATH=/tmp
PGBACKREST_REPO1_HOST_CMD=/usr/local/bin/archive-push-s3.sh
PGBACKREST_REPO1_S3_BUCKET=pgo-backups
PGBACKREST_REPO1_S3_CA_FILE=/sshd/aws-s3-ca.crt
PGBACKREST_REPO1_S3_ENDPOINT=xxx.test.com
PGBACKREST_REPO1_S3_KEY=pgo
*PGBACKREST_REPO1_S3_KEY_SECRET*=xxxxxxxxxxxxx
PGBACKREST_REPO1_S3_REGION=eu-west-1
PGBACKREST_REPO_PATH=/backrestrepo/test-backrest-shared-repo
PGBACKREST_STANZA=db
_='PGBACKREST env vars are set to:'
CONFIG is..
authorized_keys
aws-s3-ca.crt
aws-s3-credentials.yaml
config
id_rsa
ssh_host_rsa_key
sshd_config
REPO is ...
lost+found
creating  /backrestrepo/test-backrest-shared-repo
```

Not sure it makes any difference, but we're using a S3-compatible storage, not AWS S3.

**Comment by jkatz** [comment]
@dvianello Thanks for sharing, that's very helpful.

There were a few areas that I was aware of that I circled (e.g. the env vars being printed out, that should not be happening period, and that one is at least an easy fix). 

We definitely need to change this up. I'm adding it to our roadmap.

Thanks!

**Comment by jkatz** [comment]
Hi @dvianello ,

I am happy to report that I believe I have addressed all areas in #1352

Thanks again for reporting!
